src/DqtApi/DataStore/Crm/DataverseAdapter.UpdateTeacher.cs

```diff
@@ -0,0 +1,563 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using DqtApi.DataStore.Crm.Models;
+using FluentValidation;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Xrm.Sdk;
+using Microsoft.Xrm.Sdk.Messages;
+
+namespace DqtApi.DataStore.Crm
+{
+    public partial class DataverseAdapter
+    {
+        public async Task<UpdateTeacherResult> UpdateTeacher(UpdateTeacherCommand command)
+        {
+            var (result, _) = await UpdateTeacherImpl(command);
+            return result;
+        }
+
+        // Helper method that outputs the write requests that were sent for testing
+        internal async Task<(UpdateTeacherResult Result, ExecuteTransactionRequest TransactionRequest)> UpdateTeacherImpl(
+            UpdateTeacherCommand command)  // This is parameterised so we can swap out in tests
+        {
+            var helper = new UpdateTeacherHelper(this, command);
+            var referenceData = await helper.LookupReferenceData();
+            var failedReasons = helper.ValidateReferenceData(referenceData);
+            if (failedReasons != UpdateTeacherFailedReasons.None)
+            {
+                return (UpdateTeacherResult.Failed(failedReasons), null);
+            }
+
+            var lookupData = await helper.LookupData();
+            var (itt, ittLookupFailed) = helper.SelectIttRecord(lookupData.Itt, lookupData.IttProvider.Id);
+            var isEarlyYears = command.InitialTeacherTraining.ProgrammeType.IsEarlyYears();
+            if (isEarlyYears && lookupData.Teacher.dfeta_EYTSDate.HasValue)
+            {
+                return (UpdateTeacherResult.Failed(UpdateTeacherFailedReasons.AlreadyHaveEytsDate), null);
+            }
+            else if (!isEarlyYears && lookupData.Teacher.dfeta_QTSDate.HasValue)
+            {
+                return (UpdateTeacherResult.Failed(UpdateTeacherFailedReasons.AlreadyHaveQtsDate), null);
+            }
+
+            var (qualification, qualificationFailed) = helper.SelectQualificationRecord(lookupData.Qualifications);
+            if (itt != null && itt.dfeta_ProgrammeType.Value.IsEarlyYears())
+            {
+                if (command.InitialTeacherTraining.ProgrammeType != itt.dfeta_ProgrammeType)
+                {
+                    return (UpdateTeacherResult.Failed(UpdateTeacherFailedReasons.CannotChangeProgramType), null);
+                }
+            }
```

Comment:
Conditions aren't quite right here; you can change programme type you just can't change the Early Years-ness.
```suggestion
            if (itt != null && itt.dfeta_ProgrammeType.Value.IsEarlyYears() != command.InitialTeacherTraining.ProgrammeType.IsEarlyYears())
            {
                return (UpdateTeacherResult.Failed(UpdateTeacherFailedReasons.CannotChangeProgramType), null);
            }
```